pkg/workload/schemachange/generate.go

```diff
@@ -0,0 +1,223 @@
+	if atLeast < 0 || atMost < 0 || atLeast > atMost {
+		panic(errors.AssertionFailedf(
+			"bad arguments to PickBetween: [%d, %d]", atLeast, atMost,
+		))
+	}
+
+	if len(options) < atLeast {
+		return nil, errors.WithStack(ErrCaseNotPossible)
+	}
+
+	if atMost > len(options) {
+		atMost = len(options)
+	}
+
+	cloned := slices.Clone(options)
+
+	rng.Shuffle(len(cloned), func(i, j int) {
+		cloned[i], cloned[j] = cloned[j], cloned[i]
+	})
+
+	length := atLeast + rng.Intn(atMost-atLeast+1)
+	return options[:length], nil
+}
+
+// Values is a helper for formatting a list of SQL values within a
+// [GenerationCase]. It delegates to [tree.AsString] and joins the resultant
+// strings with commas.
+// Example Template Usage: `INSERT INTO table VALUES ({HelperThatReturnsValues})
+type Values []tree.NodeFormatter
+
+func AsValues[T tree.NodeFormatter](values []T, err error) (Values, error) {
+	if err != nil {
+		return nil, err
+	}
+
+	nfs := make([]tree.NodeFormatter, len(values))
+	for i := range values {
+		nfs[i] = values[i]
+	}
+	return nfs, nil
+}
+
+// String implements fmt.Stringer.
+func (v Values) String() string {
+	var b strings.Builder
+	for i, nf := range v {
+		if i > 0 {
+			b.WriteString(", ")
+		}
+		b.WriteString(tree.AsString(nf))
+	}
+	return b.String()
+}
+
+// GenerationCase is an "interesting" form of a specific [tree.Statement] and
+// the [pgcode.Code] it is expected to return.
+type GenerationCase struct {
+	// Code is the expected [pgcode.Code] of this case. Successful cases should
+	// use [pgcode.SuccessfulCompletion]
+	Code pgcode.Code
+	// Template is a text/template using `{` and `}` as delimiters that should be
+	// parsable as SQL when executed.
+	// Example: `CREATE TABLE {TableName} ({Columns})`
+	Template string
+}
+
+type compiledCase struct {
+	Code         pgcode.Code
+	TemplateName string
+}
+
+// Generate is a helper to produce a psuedo-random case for a given
+// [tree.Statement]. Functions may return ErrCaseNotPossible to indicate that
+// they can't return an appropriate value. Closures may be used to track state
+// of generation. For example, a "Table" function may set a local variable so
+// that a "Columns" function knows which table to select columns from rather
+// than buffering all possible columns into memory.
+// Usage:
+//
+//	Generate[*tree.AlterType](og.params.rng, og.produceError(), []GenerationCase{
+//		{pgcode.SuccessfulCompletion, `ALTER TYPE {Type} ADD VALUE {Value}`},
+//		{pgcode.FileAlreadyExists, `ALTER TYPE "NeverExists" ADD VALUE "Irrelevant"`},
+//	}, template.FuncMap{
+//		"Type": func() *tree.Name {
+//			// Types that implement tree.NodeFormatter should be used to properly escape values.
+//			return PickOne(og.params.rng, existingTypes)
+//		},
+//		"Value": func() string {
+//			return "foo"
+//		},
+//	})
+func Generate[T tree.Statement](
+	rng *rand.Rand, produceError bool, cases []GenerationCase, funcs template.FuncMap,
+) (T, pgcode.Code, error) {
```

Comment:
Really simple question: does the return `T, pgcode.Code, _` mean a randomly picked statement of type T such that if executed, the expected pgcode is `pgcode.Code?